### PR TITLE
templating return type of loss functions

### PR DESCRIPTION
**Pull request by mrityunjay-tripathi**
In accordance to #2338
Any other suggestions/ideas are most welcome.

### CODE REVIEW

**Comment by mrityunjay-tripathi** [comment]
Hi, Can anyone tell what has happened to Linux builds?

**Comment by zoq** [comment]
I think what you have seen is related to: https://azure.microsoft.com/en-us/blog/our-commitment-to-customers-and-microsoft-cloud-services-continuity/

File: src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp [hunk 1]
```diff
@@ -30,13 +30,15 @@ HuberLoss<InputDataType, OutputDataType>::HuberLoss(
...
 template<typename InputDataType, typename OutputDataType>
 template<typename InputType, typename TargetType>
-double HuberLoss<InputDataType, OutputDataType>::Forward(
-    const InputType& input, const TargetType& target)
+typename InputType::value_type
+HuberLoss<InputDataType, OutputDataType>::Forward(const InputType& input,
+                                                  const TargetType& target)
 {
-  double loss = 0;
+  using RealType = typename InputType::value_type;
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: In some other places we use: `typedef typename MatType::elem_type ElemType;` do you mind to change the style to be consistent with the rest of the codebase?

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: Sure, @zoq. Also ```value_type``` -> ```elem_type```?

Comment (zoq) [reply src/mlpack/methods/ann/loss_functions/huber_loss_impl.hpp]: Yes, looks like you already did that 👍 .

File: src/mlpack/methods/ann/loss_functions/sigmoid_cross_entropy_error.hpp [hunk 2]
```diff
@@ -64,8 +64,8 @@ class SigmoidCrossEntropyError
    * @param target The target vector.
    */
   template<typename InputType, typename TargetType>
-  inline double Forward(const InputType& input,
-                        const TargetType& target);
+  inline typename InputType::elem_type Forward(const InputType& input,
+                                                const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/sigmoid_cross_entropy_error.hpp]:
```suggestion
                                               const TargetType& target);
```

Minor style issue.

File: src/mlpack/methods/ann/loss_functions/reconstruction_loss.hpp [hunk 3]
```diff
@@ -49,7 +49,8 @@ class ReconstructionLoss
    * @param target The target matrix.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/reconstruction_loss.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/negative_log_likelihood.hpp [hunk 4]
```diff
@@ -48,7 +48,8 @@ class NegativeLogLikelihood
    *        between 1 and the number of classes.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/negative_log_likelihood.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/mean_squared_logarithmic_error.hpp [hunk 5]
```diff
@@ -45,7 +45,8 @@ class MeanSquaredLogarithmicError
    * @param target The target vector.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/mean_squared_logarithmic_error.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/mean_squared_error.hpp [hunk 6]
```diff
@@ -46,7 +46,8 @@ class MeanSquaredError
    * @param target The target vector.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/mean_squared_error.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/mean_bias_error.hpp [hunk 7]
```diff
@@ -45,7 +45,8 @@ class MeanBiasError
    * @param target The target vector.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/mean_bias_error.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp [hunk 8]
```diff
@@ -55,7 +55,8 @@ class LogCoshLoss
    * @param target Target data to compare with.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/log_cosh_loss.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/kl_divergence.hpp [hunk 9]
```diff
@@ -60,7 +60,8 @@ class KLDivergence
    * @param target Target data to compare with.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/kl_divergence.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/huber_loss.hpp [hunk 10]
```diff
@@ -52,7 +52,8 @@ class HuberLoss
    * @param target The target vector.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/huber_loss.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp [hunk 11]
```diff
@@ -48,7 +48,8 @@ class HingeEmbeddingLoss
    * @param target Target data to compare with.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/hinge_embedding_loss.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/earth_mover_distance.hpp [hunk 12]
```diff
@@ -45,7 +45,8 @@ class EarthMoverDistance
    * @param target The target vector.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/earth_mover_distance.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/dice_loss.hpp [hunk 13]
```diff
@@ -62,7 +62,8 @@ class DiceLoss
    * @param target The target vector.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/dice_loss.hpp]:
```suggestion
                                        const TargetType& target);
```

File: src/mlpack/methods/ann/loss_functions/cross_entropy_error.hpp [hunk 14]
```diff
@@ -49,7 +49,8 @@ class CrossEntropyError
    * @param target The target vector.
    */
   template<typename InputType, typename TargetType>
-  double Forward(const InputType& input, const TargetType& target);
+  typename InputType::elem_type Forward(const InputType& input,
+                                         const TargetType& target);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/cross_entropy_error.hpp]:
```suggestion
                                        const TargetType& target);
```

**Comment by kartikdutt18** [comment]
Hey @mrityunjay-tripathi, @favre49, @zoq  Do you think there should be a way to test this ?
i.e. We can test it for a float datatype (or any other datatype) so that we can ensure everything is perfect but more importantly also show a new contributor or someone who likes looking at mlpack code why we did this (since this aims to solve #2338).
Testing generally makes me feel safer with the code that gets merged. I hope this is okay. Let me know what you think. Thanks a lot.

**Comment by mrityunjay-tripathi** [comment]
> Hey @mrityunjay-tripathi, @favre49, @zoq Do you think there should be a way to test this?
> i.e. We can test it for a float datatype (or any other datatype) so that we can ensure everything is perfect but more importantly also show a new contributor or someone who likes looking at mlpack code why we did this (since this aims to solve #2338).
> Testing generally makes me feel safer with the code that gets merged. I hope this is okay. Let me know what you think. Thanks a lot.

Sure @kartikdutt18. Testing for the precision of all the loss functions is going to be a lot of work. But the currently added test of loss functions doesn't have even precision sufficient for ```float``` type, leave alone ```double```. Or if you are saying just to test if we are getting correct return type or not? We can check that just by using a single line: ```typeid(x).name()``` or by checking the precision of returned value by using something like ```std::numeric_limits<InputType::elem_type>::max_digits10```. The point is if the user feeds in ```float``` type matrix as ```input``` and ```target```, float value will be returned by armadillo. This seems a thing more of armadillo to me. I'm still trying to figure out what kind of test can make sense?

**Comment by kartikdutt18** [comment]
I don't think we need to add tests for all layers. A single test to show that various data types like float can be used as both input and output. Let me know what you think.

**Comment by kartikdutt18** [comment]
Hey @mrityunjay-tripathi, I looked around and I found this issue #506, So this what I think you can do. In loss function tests use arma::fmat in some cases (not all so that we preserve double as well). We need not check for precision or datatype of output. If this works correctly, which I think it does it will return the correct output. The way we will test is by using something similar to `contradiction.` Since arma::fmat shouldn't work with current loss functions but does after your PR then we can say that templating solved the issue. I hope this makes sense. @zoq, @favre49, I think there might be a better way to test this, Could you help us out with the same. Thanks a lot.

**Comment by mrityunjay-tripathi** [comment]
It's good that builds are passing :)

**Comment by favre49** [comment]
Personally, I'm not a fan of the idea of testing float on some tests and double on others - it makes the tests more convoluted and confusing, and doesn't feel like good testing practice. To me, the point of these tests is ensuring the loss functions return the correct "answer". I agree with @mrityunjay-tripathi that the datatype of these tests feels like something that we should be trusting Armadillo on.

> We can test it for a float datatype (or any other datatype) so that we can ensure everything is perfect but more importantly also show a new contributor or someone who likes looking at mlpack code why we did this

I'm not sure I understood this? Ideally the fact that this works should be obvious from documentation, not from tests

**Comment by kartikdutt18** [comment]
> I'm not sure I understood this? Ideally the fact that this works should be obvious from documentation, not from tests

Generally I feel many people learn about the codebase by looking at the tests and why certain things are implemented they way are. Since what we had worked and we are adding another feature it made sense to me test it before we merged it. 
The way I would have tested it would be:

Create a separate Test Case, say `LossFunctionTemplateTest` and added the one or two loss function objects to show they worked on float or another datatype. This is what was done for `Valid` and `Same` padding options for convolution layers. Original tests weren't changed at all. 

Maybe it's a problem with me, I consider testing to be absolutely necessary for any feature addition, but maybe it's unnecessary here. If you will feel that it is unnecessary, @mrityunjay-tripathi can use `git revert` to back to the previous state.
Thanks a lot.

**Comment by mrityunjay-tripathi** [comment]
That's okay. No test is unnecessary. Let's wait and see what other members think.

**Comment by zoq** [comment]
I like the general idea, in ensmallen, we run each test against `arma::mat` and `arma::fmat`. Here is an example: https://github.com/mlpack/ensmallen/blob/master/tests/ftml_test.cpp.

So ideally we can do the same here in mlpack, that includes everything, not only the ann methods. That said, I think it's out of scope for this particular PR. Also, we have to make sure the datatype used makes sense, like using `arma::umat` might work but the output is probably not what we want, without proper quantization.

**Comment by zoq** [comment]
So personally I would revert the change, the changes work with our main datatype right now, so it doesn't break anything.

**Comment by kartikdutt18** [comment]
> I like the general idea, in ensmallen, we run each test against arma::mat and arma::fmat. Here is an example: https://github.com/mlpack/ensmallen/blob/master/tests/ftml_test.cpp.

This looks nice. It should not be hard to extend tests for fmat as well, I think we would be able to close #1062 when methods (other than ann) also support float and are tested. Do you mind if I open up an issue for new contributors to get involved with the codebase through testing or I can create a PR as well if this seems like something that we can add.

>  like using arma::umat might work but the output is probably not what we want, without proper quantization.

Agreed, that makes sense.

**Comment by kartikdutt18** [comment]
Till then @mrityunjay-tripathi, Kindly revert the changes and we can merge this. Hopefully tests for fmat etc. will follow soon. Sorry that you and @favre49, had to deal with me. Thanks a lot.

**Comment by zoq** [comment]
Thanks again, that is a good basis we can build on.
